File: netlify/functions/shopping-list.py
"""
Netlify Function: Shopping List API
Generates shopping list from selected recipes
"""
import json
from utils.db import execute_query, format_response, handle_error
from utils.ingredients import aggregate_ingredients, round_to_practical

def handler(event, context):
    """Main handler for shopping list endpoint"""

    # Handle CORS preflight
    if event['httpMethod'] == 'OPTIONS':
        return format_response(200, {})

    try:
        if event['httpMethod'] == 'POST':
            return generate_shopping_list(event)
        else:
            return format_response(405, {
                "error": "Method not allowed",
                "success": False
            })

    except Exception as e:
        return handle_error(e)

def generate_shopping_list(event):
    """Generate shopping list from recipe IDs"""
    try:
        # Parse request body
        body = json.loads(event['body'])

        recipe_ids = body.get('recipe_ids', [])
        exclude_pantry = body.get('exclude_pantry', False)

        if not recipe_ids:
            return format_response(400, {
                "error": "No recipes selected",
                "success": False
            })

        # Fetch selected recipes
        # Use ANY with array to match recipe IDs
        query = "SELECT * FROM recipes WHERE id = ANY(%s)"
        recipes = execute_query(query, (recipe_ids,), fetch='all')

        if not recipes:
            return format_response(404, {
                "error": "No recipes found",
                "success": False
            })

        # Aggregate ingredients
        categorized_ingredients = aggregate_ingredients(recipes, exclude_pantry)

        # Round quantities to practical amounts
        for category, ingredients in categorized_ingredients.items():
            for ing in ingredients:
                ing['quantite'] = round_to_practical(ing['quantite'], ing['unite'])

        # Calculate statistics
        total_items = sum(len(items) for items in categorized_ingredients.values())

        return format_response(200, {
            "success": True,
            "shopping_list": categorized_ingredients,
            "total_items": total_items,
            "recipe_count": len(recipes),
            "recipes": [{"id": r['id'], "titre": r['titre']} for r in recipes]
        })

    except json.JSONDecodeError:
        return format_response(400, {
            "error": "Invalid JSON in request body",
            "success": False
        })
    except Exception as e:
        return handle_error(e)

File: netlify/functions/shopping-list.mjs
import { json, err } from "./shared/http.mjs";
import { queryAll } from "./shared/db.mjs";
import { aggregateIngredients, roundToPractical } from "./shared/ingredients.mjs";

export const handler = async (event) => {
  if (event.httpMethod === "OPTIONS") return json(200, {});

  try {
    if (event.httpMethod !== "POST") {
      return err(405, "Method not allowed");
    }

    let body;
    try {
      body = JSON.parse(event.body || "{}");
    } catch {
      return err(400, "Invalid JSON in request body");
    }

    const recipeIds = body.recipe_ids || [];
    const excludePantry = Boolean(body.exclude_pantry);

    if (!recipeIds.length) return err(400, "No recipes selected");
    if (!Array.isArray(recipeIds)) return err(400, "recipe_ids must be an array");
    if (recipeIds.length > 50) {
      return err(400, "Maximum 50 recipes allowed per shopping list");
    }

    let ids;
    try {
      ids = recipeIds.map((rid) => parseInt(rid, 10));
      if (ids.some((n) => Number.isNaN(n))) throw new Error("nan");
    } catch {
      return err(400, "Invalid recipe ID format");
    }

    const placeholders = ids.map((_, i) => `$${i + 1}`).join(", ");
    const sql = `SELECT * FROM recipes WHERE id IN (${placeholders})`;
    const recipes = await queryAll(sql, ids);

    if (!recipes.length) return err(404, "No recipes found");

    let categorized = aggregateIngredients(recipes, excludePantry);
    categorized = Object.fromEntries(
      Object.entries(categorized).map(([cat, items]) => [
        cat,
        items.map((ing) => ({
          ...ing,
          quantite: roundToPractical(ing.quantite, ing.unite),
        })),
      ])
    );

    const totalItems = Object.values(categorized).reduce(
      (sum, items) => sum + items.length,
      0
    );

    return json(200, {
      success: true,
      shopping_list: categorized,
      total_items: totalItems,
      recipe_count: recipes.length,
      recipes: recipes.map((r) => ({ id: r.id, titre: r.titre })),
    });
  } catch (e) {
    console.error(e);
    return json(500, { success: false, error: "An internal error occurred" });
  }
};
